internal_transfers/actions/tests/database.spec.ts

```diff
@@ -151,10 +154,128 @@ describe("test database insertion methods", () => {
     ];
     expect(results).toStrictEqual(expected);
 
-    // Idempotency
-    await insertSettlementSimulations(db, dummySimData);
-    results = await db.query(sql`SELECT * from settlement_simulations;`);
-    expect(results).toStrictEqual(expected);
+    await expect(insertSettlementSimulations(db, dummySimData)).rejects.toThrow(
+      'duplicate key value violates unique constraint "settlement_simulations_pkey"'
+    );
+  });
+  test("insertPipelineResults", async () => {
+    const txHash = "0x4321";
+    const solver = "0xc9ec550bea1c64d779124b23a26292cc223327b6";
+    const eventMeta = { txHash: txHash, blockNumber: 0 };
+    const settlementEvent = { solver: solver, logIndex: 0 };
+
+    const token1 = "0x12";
+    const token2 = "0x34";
+    const imbalances = [
+      {
+        token: token1,
+        amount:
+          115792089237316195423570985008687907853269984665640564039457584007913129639935n,
+      },
+      { token: token2, amount: 5678n },
+    ];
+    const blockNumber = 1;
+    const gasUsed = 2;
+    const simulationID = "sim-id";
+    const largeBigInt = 99999999999999999999999999999999999999999999999999n;
+    const exampleLog = {
+      address: ethers.ZeroAddress,
+      // The indexed topics from the event log
+      topics: [ethers.ZeroHash, ethers.ZeroHash],
+      data: ethers.ZeroHash,
+    };
+    const exampleEthDelta = new Map([["0x2", largeBigInt]]);
+    const simulationDatum: SettlementSimulationData = {
+      txHash,
+      full: {
+        simulationID,
+        gasUsed,
+        blockNumber,
+        // Note there are 2 logs here!
+        logs: [exampleLog, exampleLog],
+        ethDelta: exampleEthDelta,
+      },
+      reduced: {
+        simulationID,
+        gasUsed,
+        blockNumber,
+        // and only 1 log here!
+        logs: [exampleLog],
+        ethDelta: exampleEthDelta,
+      },
+      winningSettlement: {
+        solver: ethers.ZeroAddress,
+        simulationBlock: blockNumber,
+        reducedCallData: "0xca11da7a",
+        fullCallData: "0xca11da7a000000",
+      },
+    };
+    // insertion works.
+    await insertPipelineResults(
+      db,
+      simulationDatum,
+      imbalances,
+      eventMeta,
+      settlementEvent
+    );
+    const recordCountQuery = sql`select count(*)
+                from internalized_imbalances i
+                       join settlements s
+                            on i.tx_hash = s.tx_hash
+                       join settlement_simulations ss
+                            on i.tx_hash = ss.tx_hash`;
+    let results = await db.query(recordCountQuery);
+    expect(results.length).toEqual(1);
+    expect(results[0]).toEqual({ count: 2n });
+
+    // Idempotency & insertion works ONLY IF ALL three insertions pass!
```

Comment:
My understanding is that the function `insertPipelineResults` isn't idempotent because if you run it again with the same input it errors instead of succeeding. Am I missing something?

Reply:
That is reasonable. So then I will remove the term "idempotent" here unless you think the operation should actually just act that way. 

For program efficiency, my opinion is that its better to check if an operation has already been performed (to avoid redundant simulation costs etc...), so what will actually become idempotent is the method which calls this and implements some sort of try/catch to handle the insertion failure. 
I just wanted to make sure we were informed of such situations instead of failing silently.

Reply:
Removed the term.